File: Meta/173.binary-search-tree-iterator.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class BSTIterator {
    vector<TreeNode*> st;

    void pushLeft(TreeNode* node) {
        while (node) {
            st.push_back(node);
            node = node->left;
        }
    }

public:
    BSTIterator(TreeNode* root) {
        /*
        Approach: iterative inorder traversal. The stack stores the path to the
        next smallest node. next() pops one node and pushes the left spine of its
        right subtree.

        C++ notes: vector<TreeNode*> is used as a stack of raw tree pointers.
        Complexity: next and hasNext are O(1) amortized time, O(h) space.
        */
        pushLeft(root);
    }

    int next() {
        TreeNode* node = st.back();
        st.pop_back();
        pushLeft(node->right);
        return node->val;
    }

    bool hasNext() { return !st.empty(); }
};

File: SoFi/173.binary-search-tree-iterator.go
package leetcode

// BSTIterator173 stores a stack of the path to the next inorder node. pushLeft
// descends left children; Next pops one node and then pushes its right subtree's
// left spine.
type BSTIterator173 struct{ stack []*TreeNode }

func Constructor173(root *TreeNode) BSTIterator173 {
	it := BSTIterator173{}
	it.pushLeft(root)
	return it
}
func (it *BSTIterator173) pushLeft(node *TreeNode) {
	for node != nil {
		it.stack = append(it.stack, node)
		node = node.Left
	}
}
func (it *BSTIterator173) Next() int {
	node := it.stack[len(it.stack)-1]
	it.stack = it.stack[:len(it.stack)-1]
	it.pushLeft(node.Right)
	return node.Val
}
func (it *BSTIterator173) HasNext() bool { return len(it.stack) > 0 }
